{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyOiTQtHFn854gsyY0mrLxsa"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "id": "SOhd4ibIanFT"
      },
      "outputs": [],
      "source": [
        "import numpy as np\n",
        "import random\n",
        "import matplotlib.pyplot as plt\n",
        "import pandas as pd\n",
        "import seaborn as sns"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def generate_nb(state, E=1):\n",
        "    '''\n",
        "    Given the state, generate the NB distribution\n",
        "    state: int\n",
        "    E: int. Gene expression level (1,5,10,15,20)\n",
        "    '''\n",
        "    mean_list_true = np.array([60, 350, 300, 250, 200, 150, 120, 320, 250, 220, 60, 450, 400, 300, 250, 230, 100, 430, 350, 320, 60]) # mean\n",
        "    variance_list_true = mean_list_true*1.2 \n",
        "    beta_list_true = 1/(variance_list_true/mean_list_true-1)\n",
        "    alpha_list_true = mean_list_true * beta_list_true\n",
        "    return np.random.negative_binomial(alpha_list_true[state-1], beta_list_true[state-1]/(E+beta_list_true[state-1]))"
      ],
      "metadata": {
        "id": "aBwFRiCHAbJA"
      },
      "execution_count": 20,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def simulate_single():\n",
        "    '''\n",
        "    Simulate single RNA sequence contains 2 scenarios: both uORF and CDS, only CDS\n",
        "    start codon: AUG\n",
        "    stop codons: UAA, UGA, UAG\n",
        "    '''\n",
        "    stop_codon_list = ['UAA','UGA','UAG']\n",
        "\n",
        "    # indexs for initiation\n",
        "    state_1_2 = 0\n",
        "    state_1_12 = 0\n",
        "    state_11_12 = 0\n",
        "\n",
        "    # first several bases without start codons\n",
        "    curr_RNA = random.choices(['A', 'C', 'G', 'U'], weights=np.ones(4), k = 5)\n",
        "    while 'AUG' in \"\".join(curr_RNA):\n",
        "        curr_RNA = random.choices(['A', 'C', 'G', 'U'], weights=np.ones(4), k = 5)\n",
        "    observed_counts = [generate_nb(1), generate_nb(1), generate_nb(1), generate_nb(1), generate_nb(1)]\n",
        "    states_seq = [1, 1, 1, 1, 1]\n",
        "    state = 1\n",
        "\n",
        "\n",
        "    # 5'U\n",
        "    while state == 1: \n",
        "      codons = random.choices(['A', 'C', 'G', 'U'], weights=np.ones(4), k = 3)\n",
        "      if \"\".join(codons) == 'AUG':\n",
        "        codons = random.choices(['A', 'C', 'G', 'U'], weights=np.ones(4), k = 3)\n",
        "      curr_RNA += codons\n",
        "      observed_counts.extend((generate_nb(1),generate_nb(1),generate_nb(1)))\n",
        "      states_seq.extend((1,1,1))\n",
        "      state = random.choices([1, 2, 12], weights=[0.2, 0.5, 0.3], k = 1)[0]\n",
        "\n",
        "    if state == 2:\n",
        "      state_1_2 = 1\n",
        "    elif state == 12:\n",
        "      state_1_12 = 1\n",
        "\n",
        "    # uORF\n",
        "    if state == 2:\n",
        "\n",
        "      # initiation site\n",
        "      curr_RNA += ['A', 'U', 'G'] # add start codon\n",
        "      observed_counts.extend((generate_nb(2),generate_nb(3),generate_nb(4)))\n",
        "      states_seq.extend((2,3,4))\n",
        "\n",
        "      # elongation recycing: set shorter elongation compared to CDS \n",
        "      repeat = random.choices(np.arange(10, 31), weights=np.ones(21), k = 1)[0]\n",
        "      for i in range(repeat):\n",
        "        codons = random.sample(['A', 'C', 'G', 'U'], 3)\n",
        "        if \"\".join(codons) in stop_codon_list:\n",
        "          codons = random.sample(['A', 'C', 'G', 'U'], 3)\n",
        "        curr_RNA += codons\n",
        "        observed_counts.extend((generate_nb(5),generate_nb(6),generate_nb(7)))\n",
        "        states_seq.extend((5,6,7))\n",
        "\n",
        "      # termination site\n",
        "      curr_RNA += random.choices([['U', 'A', 'A'], ['U', 'G', 'A'], ['U', 'A', 'G']], weights=[1, 1, 1], k = 1)[0]\n",
        "      observed_counts.extend((generate_nb(8),generate_nb(9),generate_nb(10)))\n",
        "      states_seq.extend((8,9,10))\n",
        "\n",
        "      # 5'U2\n",
        "      state = 11\n",
        "      temp = random.choices(['A', 'C', 'G', 'U'], weights=np.ones(4), k = 10)\n",
        "      while 'AUG' in \"\".join(temp):\n",
        "          temp = random.choices(['A', 'C', 'G', 'U'], weights=np.ones(4), k = 10)\n",
        "      curr_RNA += temp\n",
        "      observed_counts.extend((generate_nb(11),generate_nb(11),generate_nb(11),generate_nb(11),generate_nb(11),generate_nb(11),generate_nb(11),generate_nb(11),generate_nb(11),generate_nb(11)))\n",
        "      states_seq.extend((11,11,11,11,11,11,11,11,11,11))\n",
        "\n",
        "      while state == 11:\n",
        "        codons = random.choices(['A', 'C', 'G', 'U'], weights=np.ones(4), k = 3)\n",
        "        if \"\".join(codons) == 'AUG':\n",
        "          codons = random.choices(['A', 'C', 'G', 'U'], weights=np.ones(4), k = 3)\n",
        "        curr_RNA += codons\n",
        "        observed_counts.extend((generate_nb(11),generate_nb(11),generate_nb(11)))\n",
        "        states_seq.extend((11,11,11))\n",
        "        #if len(curr_RNA) >= 100:\n",
        "           #return [curr_RNA, observed_counts, states_seq, [state_1_2, state_1_12, state_11_12]]\n",
        "        state = random.choices([11, 12], weights=[0.2, 0.8], k = 1)[0]\n",
        "        if state == 12:\n",
        "          state_11_12 = 1\n",
        "\n",
        "   # CDS\n",
        "    if state == 12:\n",
        "\n",
        "     # initiation site\n",
        "      curr_RNA += ['A', 'U', 'G'] # add start codon\n",
        "      observed_counts.extend((generate_nb(12),generate_nb(13),generate_nb(14)))\n",
        "      states_seq.extend((12,13,14))\n",
        "\n",
        "     # elongation recycing: set longer elongation compared to uORF \n",
        "      repeat = random.choices(np.arange(32, 53), weights=np.ones(21), k = 3)[0]\n",
        "      for i in range(repeat):\n",
        "        codons = random.choices(['A', 'C', 'G', 'U'], weights=np.ones(4), k = 3)\n",
        "        if \"\".join(codons) in stop_codon_list:\n",
        "          codons = random.choices(['A', 'C', 'G', 'U'], weights=np.ones(4), k = 3)\n",
        "        curr_RNA += codons\n",
        "        observed_counts.extend((generate_nb(15),generate_nb(16),generate_nb(17)))\n",
        "        states_seq.extend((15,16,17))\n",
        "\n",
        "     # termination site\n",
        "      curr_RNA += random.choices([['U', 'A', 'A'], ['U', 'G', 'A'], ['U', 'A', 'G']], weights=[1, 1, 1], k = 1)[0]\n",
        "      observed_counts.extend((generate_nb(18),generate_nb(19),generate_nb(20)))\n",
        "      states_seq.extend((18,19,20))  \n",
        "\n",
        "     # 3'U\n",
        "      repeat = random.choices(np.arange(10, 31), weights=np.ones(21), k = 1)[0]\n",
        "      for i in range(repeat):\n",
        "        curr_RNA += random.choices(['A', 'C', 'G', 'U'], weights=np.ones(4), k = 3)\n",
        "        observed_counts.extend((generate_nb(21),generate_nb(21),generate_nb(21)))\n",
        "        states_seq.extend((21,21,21))\n",
        "\n",
        "    return [curr_RNA, observed_counts, states_seq, [state_1_2, state_1_12, state_11_12]]\n"
      ],
      "metadata": {
        "id": "-0DahpDLiOEb"
      },
      "execution_count": 260,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def simulate_main(length):\n",
        "    '''\n",
        "    Simulate multiple RNA sequence contains 3 scenarios: only uORF + only CDS + both\n",
        "    length: the number of RNA sequences\n",
        "\n",
        "    start codon: AUG\n",
        "    stop codons: UAA, UGA, UAG\n",
        "    ''' \n",
        "\n",
        "    RNA_data = []\n",
        "    counts_data = []\n",
        "    states_data = []\n",
        "    transition_list = []\n",
        "\n",
        "    for i in range(length):\n",
        "        temp = simulate_both()\n",
        "        \n",
        "        # change to only uORF\n",
        "        if i%3 == 0:\n",
        "          if temp[3][0] == 1 and temp[3][2] == 1:\n",
        "            RNA_seq = temp[0]\n",
        "            observed_counts = temp[1]\n",
        "            states_seq = temp[2]\n",
        "           \n",
        "            start_index = states_seq.index(12)\n",
        "            stop_index = states_seq.index(20)\n",
        "            final_index = states_seq.index(21)\n",
        "\n",
        "            # replace original state sequence\n",
        "            states_replace = [11 for i in states_seq[start_index:]]\n",
        "            states_seq[start_index:] = states_replace\n",
        "\n",
        "            # replace original count sequence\n",
        "            count_replace = [generate_nb(11) for i in observed_counts[start_index:]]\n",
        "            observed_counts[start_index:] = count_replace\n",
        "\n",
        "            # replace original RNA sequence\n",
        "            for j in range(start_index,stop_index-2,3):\n",
        "              condons_replace = random.choices(['A','C', 'G', 'U'], weights=np.ones(4), k = 3)\n",
        "              if \"\".join(condons_replace) == 'AUG':\n",
        "                  condons_replace = random.choices(['A','C', 'G', 'U'], weights=np.ones(4), k = 3)\n",
        "              RNA_seq[j:(j+3)] = condons_replace\n",
        "\n",
        "            condons_replace1 = [random.choices(['A','C', 'G', 'U'], weights=np.ones(4), k = 1)[0] for r in RNA_seq[final_index:]]\n",
        "            RNA_seq[final_index:] = condons_replace1\n",
        "\n",
        "            temp[0] = RNA_seq\n",
        "            temp[1] = observed_counts\n",
        "            temp[2] = states_seq\n",
        "            temp[3][2] = 0\n",
        "\n",
        "            RNA_data.append(temp[0])\n",
        "            counts_data.append(temp[1])\n",
        "            states_data.append(temp[2])\n",
        "            transition_list.append(temp[3])\n",
        "        \n",
        "        RNA_data.append(temp[0])\n",
        "        counts_data.append(temp[1])\n",
        "        states_data.append(temp[2])\n",
        "        transition_list.append(temp[3])\n",
        "  \n",
        "    return [RNA_data, counts_data, states_data, transition_list]"
      ],
      "metadata": {
        "id": "V7Kaq_V4_HQZ"
      },
      "execution_count": 261,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def simulate_neither(length):\n",
        "  '''\n",
        "  Simulate multiple RNA sequence without translated regions\n",
        "  length: the number of RNA sequences\n",
        "  ''' \n",
        "  RNA_data = []\n",
        "  counts_data = []\n",
        "  states_data = []\n",
        "  transition_list = []\n",
        "  #RNA_data = random.choices(['A', 'C', 'G', 'U'], weights=np.ones(4), k = 5)\n",
        "  #while 'AUG' in \"\".join(RNA_data):\n",
        "    #RNA_data = random.choices(['A', 'C', 'G', 'U'], weights=np.ones(4), k = 5)\n",
        "  #counts_data = [generate_nb(1), generate_nb(1), generate_nb(1), generate_nb(1), generate_nb(1)]\n",
        "  #states_data = [1, 1, 1, 1, 1]\n",
        "  #transition_list = []\n",
        "\n",
        "  for i in range(length):\n",
        "    RNA_seq = []\n",
        "    observed_counts = []\n",
        "    states_seq = []\n",
        "    repeat = random.choices(np.arange(30, 81), weights=np.ones(51), k = 1)[0]\n",
        "    for j in range(repeat):\n",
        "        codons = random.choices(['A', 'C', 'G', 'U'], weights=np.ones(4), k = 3)\n",
        "        RNA_seq.extend(codons)\n",
        "        observed_counts.extend((generate_nb(1),generate_nb(1),generate_nb(1)))\n",
        "        states_seq.extend((1,1,1))\n",
        "\n",
        "    RNA_data.append(RNA_seq)\n",
        "    counts_data.append(observed_counts)\n",
        "    states_data.append(states_seq)\n",
        "    transition_list.append([0,0,0])\n",
        "    \n",
        "  return [RNA_data, counts_data, states_data, transition_list]\n"
      ],
      "metadata": {
        "id": "AQ_wfXoMBUzf"
      },
      "execution_count": 193,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# generate datasets\n",
        "random.seed(0)\n",
        "part1 = simulate_main(2700) \n",
        "part2 = simulate_neither(300)  \n",
        "RNA_data = part1[0] + part2[0]\n",
        "observed_data = part1[1] + part2[1]\n",
        "states_seq = part1[2] + part2[2]\n",
        "transition_list = part1[3] + part2[3]"
      ],
      "metadata": {
        "id": "V5HBoSB1L74N"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# save as file: RNA_data, counts_data, states_data and transition_list\n",
        "def saveListToFile(listname, pathtosave):\n",
        "    file1 = open(pathtosave,\"w\") \n",
        "    for i in listname:\n",
        "        file1.writelines(\"{}\\n\".format(i))    \n",
        "    file1.close() \n",
        "\n",
        "listname = [RNA_data, observed_data, states_seq, transition_list]\n",
        "saveListToFile(listname, path)"
      ],
      "metadata": {
        "id": "TI9e9pZ8L9u_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# visualization\n",
        "# both\n",
        "# x_tick = list(map(lambda num: \"\" if num % 10 != 0 else num, range(len(observed_data[2]))))\n",
        "# plt.figure(figsize=(300, 100), dpi=10)\n",
        "# plt.ylabel('Number of RPF counts')\n",
        "# plt.grid()\n",
        "# plt.bar(np.arange(len(observed_data[2])), observed_data[2], width=0.5)\n",
        "# plt.ylim(0,500)\n",
        "# plt.xlim(0,350)\n",
        "# plt.xticks(range(len(x_tick)), x_tick, size=200)\n",
        "# plt.yticks(size=200)\n",
        "# plt.show()\n",
        "# # only main coding region\n",
        "# plt.bar(np.arange(len(observed_data[0])), observed_data[0])\n",
        "# plt.ylim(0,500)\n",
        "# plt.xlim(0,350)\n",
        "# plt.show()\n",
        "# # only uORF\n",
        "# plt.bar(np.arange(len(observed_data[5])), observed_data[5])\n",
        "# plt.ylim(0,500)\n",
        "# plt.xlim(0,350)\n",
        "# plt.show()\n",
        "# # none\n",
        "# plt.bar(np.arange(len(observed_data[2500])), observed_data[2500])\n",
        "# plt.ylim(0,500)\n",
        "# plt.xlim(0,350)\n",
        "# plt.show()"
      ],
      "metadata": {
        "id": "rdyKD80xKUU0"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}